src/planner.c

```diff
@@ -796,6 +796,27 @@ timescaledb_get_relation_info_hook(PlannerInfo *root, Oid relation_objectid, boo
 	switch (classify_relation(root, rel, &ht))
 	{
 		case TS_REL_HYPERTABLE:
+		{
+#if PG12_GE
+			/* This only works for PG12 because for earlier versions the inheritance
+			 * expansion happens too early during the planning phase
+			 */
+			RangeTblEntry *rte = planner_rt_fetch(rel->relid, root);
+			Query *query = root->parse;
+			/* Mark hypertable RTEs we'd like to expand ourselves
+			 * Hypertables inside inlineable functions don't get marked during the query
+			 * preprocessing step. Therefore we do an extra try here. However, we need to
+			 * be careful for UPDATE/DELETE as Postgres runs them through the planner as
+			 * a simulated SELECT. We don't want to touch these so we have to check
+			 * rte->requiredPerms if we're requesting an UPDATE/DELETE on this rel.
+			 */
+			if (!ts_guc_disable_optimizations && ts_guc_enable_constraint_exclusion && inhparent &&
+				rte->ctename == NULL && !IS_UPDL_CMD(query) && query->resultRelation == 0 &&
```

Comment:
It happens in `planner.c` in Postgres source, around line 1300:

```
		/*
		 * Make a deep copy of the parsetree for this planning cycle to mess
		 * around with, and change it to look like a SELECT.  (Hack alert: the
		 * target RTE still has updatedCols set if this is an UPDATE, so that
		 * expand_partitioned_rtentry will correctly update
		 * subroot->partColsUpdated.)
		 */
		subroot->parse = copyObject(root->parse);

		subroot->parse->commandType = CMD_SELECT;
		subroot->parse->resultRelation = 0;
```

Indeed that makes `!IS_UPDL_CMD(query)` check redundant, because that'll always evaluate to true on v12. However, since I'd say Postgres applies a bit of a hack for planning update/delete's in v12, I think it's good to check anyway and not rely on the Postgres hack completely. If you disagree I'd be happy to remove it though.

Queries that trigger this behavior can actually be found in TimescaleDb tests already. For example, when running the tests without the extra ACL permission check, regressions come up. Here's one for an `UPDATE`, but a similar test exists for `DELETE`.

```
-- ConstraintAwareAppend NOT applied for UPDATE
EXPLAIN (costs off)
UPDATE "one_Partition"
SET series_1 = 8
WHERE series_1 IN (SELECT series_1 FROM "one_Partition" WHERE series_1 > series_val());
                                                                   QUERY PLAN                                                                   
------------------------------------------------------------------------------------------------------------------------------------------------
 Update on _hyper_1_1_chunk <<<<------- incorrect
   Update on _hyper_1_1_chunk
   Update on _hyper_1_2_chunk
   Update on _hyper_1_3_chunk
   ->  Hash Join
         Hash Cond: (_hyper_1_1_chunk.series_1 = "one_Partition".series_1)
         ->  Seq Scan on _hyper_1_1_chunk
         ->  Hash
               ->  HashAggregate
                     Group Key: "one_Partition".series_1
etc...
```

Reply:
After some experimenting it would seem I cannot move them from the if-statement to the assert unfortunately.
The Postgres planner code I gave you is only part of the full, rather complicated, inheritance planner. The first step of the planner is what I showed - where it simulates it as a `SELECT` to find out which partitions to process. However, it then uses the results of this to modify its own `UPDATE`/`DELETE` plan and then runs it through the planner again. This time, it actually *resets* the `requiredPerms` field to `0` though.
This means that:
- In the first part of the planner, `commandType==SELECT` and `requiredPerms` are set to `UPDATE/DELETE`
- In the second part, it's called again but now with `commandType==UPDATE/DELETE` and `requiredPerms==0`.
The reset happens in `inherit.c` function `expand_single_inheritance_child`

These points together means we do need to check all conditions in the if-statement.
I will add some extra comments to this part though.

Reply:
Ok, then they need to be kept, but please add tests that make sure that nobody accidentally removes them and triggers a regression. That is, a test that will fail if any of the checks are removed, but not otherwise.

Reply:
Yes, I'm adding extra tests now. Will post a new version soon.